x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativeRolesStore.java

```diff
@@ -186,15 +199,25 @@ public void getRoleDescriptors(Set<String> names, final ActionListener<RoleRetri
             return;
         }
 
+        final Set<String> rolesToGet = filterReservedRoleNames(names);
```

Comment:
Filtering all roles here to avoid that reserved roles get resolved by `NativeRolesStore` when `getRoleDescriptors` is called from `accept` method to resolve roles and later used during authorization.

Reply:
Hmm I'm not sure that's necessary: 

We [remove](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/RoleDescriptorStore.java#L300-L303) already resolved roles out before handing off to the next resolver in the chain.

So if we are resolving e.g. `viewer` in stateful, it'll get returned by the reserved role store, and won't get checked by the native role store. Similarly, the file store will come first so any role that gets resolved there won't get resolved by the native role store.

We could make this an assertion here (i.e., no reserved roles ever make it to this point).

Reply:
Nice catch! I've misted the fact that resolved are removed. It makes sense to change it to an assertion.

Reply:
The filtering is still needed for Serverless, which was the main reason I've added it.
I've introduced it back in the [`TransportGetRolesAction`](https://github.com/elastic/elasticsearch/pull/121971/files#diff-b3d9b568891fc844f44739495b9a3551f100a0d8d99ebd1222ed5ffdcab59b95R56-R61). I think this is okay given that we already do filtering of reserved roles there. The main change is using the `ReservedRoleNameChecker` instead of `ReservedRolesStore`.

Reply:
Would you mind taking another look?

Reply:
Yikes, sorry about that. You're totally right -- we hit a different path when fetching by name. Adding the check sounds good to me. Taking a quick look now 👀